R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+#'   data. Default is 5 years.
+#' @param t2 The number of years into the future against which you want to
+#'   calculate disruption. Default is 10 years.
+#' @param additional_groups Character vector. The names of columns to group by,
+#'   in addition to these ones (which are always used):
+#'
+#'   `r toString(crucial_tdm_groups())`
+#'
+#' @return A tibble with the columns specified in the `...` input, as well as:
+#'   `tdm_tech`, the technology level transition disruption metric and
+#'   `tdm_sec`, the sector level transition disruption metric.
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#'
+#' pacta_results <- tibble(
+#'   investor_name = rep("some_investor", 3),
+#'   portfolio_name = rep("some_portfolio", 3),
+#'   allocation = rep("portfolio_weight", 3),
+#'   equity_market = rep("GlobalMarket", 3),
+#'   scenario_geography = rep("Global", 3),
+#'   ald_sector = rep("Power", 3),
+#'   technology = rep("RenewablesCap", 3),
+#'   year = c(2020, 2025, 2030),
+#'   plan_alloc_wt_tech_prod = rep(1, 3),
+#'   scen_alloc_wt_tech_prod = 1:3,
+#'   plan_carsten = rep(0.5, 3)
+#' )
+#' pacta_results
+#'
+#' calculate_tdm(pacta_results, t0 = 2020)
+calculate_tdm <- function(data, t0, t1 = 5, t2 = 10, additional_groups = NULL) {
+  stopifnot(
+    is.data.frame(data),
+    is.numeric(t0),
+    is.numeric(t1),
+    is.numeric(t2)
+  )
+
+  if (!is.null(additional_groups)) stopifnot(is.character(additional_groups))
+
+  groups <- union(crucial_tdm_groups(), additional_groups)
+
+  check_calculate_tdm(data, t0, t1, t2, groups)
+
+  filtered_data <- filter(data, .data$allocation == "portfolio_weight")
+
+  if (nrow(filtered_data) == 0) {
+    return(warn_zero_rows(tdm_prototype()))
+  }
+
+  # TODO: Filter the data fora  particular scenario. Function should gain the
+  # argument `scenario`, which by default specifies IPR - FPS scenario. I am
+  # waiting until the data is prepared as I'm not sure exactly what this
+  # scenario will be called in the input data.
+
+  data_with_monotonic_factors <- add_monotonic_factor(filtered_data, t0, t1, t2, groups)
+
+  initial_year_data <- filter(data_with_monotonic_factors, .data$year == t0)
+
+  preformatted_data <- pre_format_data(data_with_monotonic_factors, t0, t1, t2, groups)
+
+  data_with_tdm <- add_tdm(preformatted_data, groups)
+
+  formatted_data_with_tdm <- left_join(
+    initial_year_data,
+    data_with_tdm,
+    by = groups
+  )
+
+  formatted_data_with_tdm %>%
+    add_aggregate_tdm(t0, groups) %>%
+    select(names(tdm_prototype()), all_of(groups))
+}
+
+warn_zero_rows <- function(data) {
+  # NOTE: This function will only work if the allocation method is
+  # portfolio_weight. ownership_weight outputs 0 for all carsten metric values,
+  # and thus we would be dividing by zero otherwise...
+  message <- 'Filtering for "portfolio_weight" allocation, outputs 0 rows'
+  warn(message, class = "has_zero_rows")
+
+  # Pass `data` to inline inside return()
+  invisible(data)
```

Comment:
I don't understand what this means. I think the best way to explain `invisible(data)` is to allow using this function in a pipe. But then it's a well documented design pattern so I don't think you need to explain it.

https://design.tidyverse.org/out-invisible.html

Reply:
hahaha i think this was actually something you added in a commit!! maybe I'm wrong, but I have no idea what is meant by that comment.